src/lib.rs

```diff
@@ -621,182 +654,203 @@ where
                 _ => continue,
             }
 
-            // Where does this function live?
-            let ranges = Self::get_ranges(entry, debug_ranges, header.address_size(), base_address)
-                .chain_err(|| "subroutine has invalid ranges")?;
-            if ranges.is_empty() {
-                continue;
+            if let Some(program) = self.parse_program(entry)? {
+                programs.push(program);
             }
+        }
+        self.programs = programs;
+        self.read_programs = true;
+        Ok(())
+    }
+
+    fn parse_program<'a, 'b>(
+        &self,
+        entry: &gimli::DebuggingInformationEntry<'a, 'b, gimli::EndianBuf<'input, Endian>>,
+    ) -> Result<Option<Program<'input, Endian>>> {
+        let mut name = None;
+        let mut abstract_origin = None;
+        let mut specification = None;
+        let mut low_pc = None;
+        let mut high_pc = None;
+        let mut size = None;
+        let mut range_offset = None;
+
+        let mut attrs = entry.attrs();
+        while let Some(attr) = attrs
+            .next()
+            .map_err(|e| Error::from(ErrorKind::Gimli(e)))
+            .chain_err(|| "invalid subprogram attribute")?
+        {
+            match attr.name() {
```

Comment:
We do essentially this match for the compilation unit too. Maybe it should be factored out?

Reply:
It's hard to factor out because here there's some extra attributes that are needed (`low_pc`/`high_pc`/`ranges`).